src/librustc_resolve/build_reduced_graph.rs

```diff
@@ -99,154 +99,259 @@ impl<'a> Resolver<'a> {
+                        if let ast::UseTreeKind::Simple(..) = use_tree.kind {
+                            if use_tree.ident().name == keywords::SelfValue.name() {
+                                return Some(use_tree.span);
+                            }
+                        }
+
+                        None
+                    }).collect::<Vec<_>>();
+                    if self_spans.len() > 1 {
+                        let mut e = resolve_struct_error(self,
+                            self_spans[0],
+                            ResolutionError::SelfImportCanOnlyAppearOnceInTheList);
 
-                    for other_span in self_spans.iter().skip(1) {
-                        e.span_label(*other_span, "another `self` import appears here");
+                        for other_span in self_spans.iter().skip(1) {
+                            e.span_label(*other_span, "another `self` import appears here");
+                        }
+
+                        e.emit();
                     }
 
-                    e.emit();
-                }
+                    // `use x::y::{self, ...}` has the same ambiguity-catching effect as
+                    // `forbid_ambiguous_prefix`, so at most one should be present,
+                    // otherwise ambiguity errors would appear to be duplicated.
+                    if self_spans.is_empty() {
+                        forbid_ambiguous_prefix(self, &module_path);
+                    }
 
-                for &(ref tree, id) in items {
-                    self.build_reduced_graph_for_use_tree(
-                        root_use_tree, root_id, tree, id, vis, &prefix, true, item, expansion
-                    );
+                    for &(ref tree, id) in items {
+                        self.build_reduced_graph_for_use_tree(
+                            root_use_tree,
+                            root_id,
+                            tree,
+                            id,
+                            vis,
+                            &prefix,
+                            prefix_is_half_of_extern_and_relative_duo || include_relative_import,
+                            true,
+                            item,
+                            expansion,
+                        );
+                    }
                 }
             }
-        }
+        };
+
+        if include_relative_import {
+            // Desugar from `use ident::suffix;` into
+            // `use self::ident::suffix;` and
+            // `use extern::ident::suffix;`
+            let mut self_module_path = module_path.clone();
+            let mut extern_module_path = module_path;
+            let self_ident = Ident { name: keywords::SelfValue.name(), span: use_tree.span };
+            let extern_ident = Ident { name: keywords::Extern.name(), span: use_tree.span };
+            self_module_path[0] = self_ident;
+            extern_module_path[0] = extern_ident;
```

Comment:
This doesn't need to do anything, since `CrateRoot` and `Extern` have the same effect.